knative-operator/pkg/common/telemetry/telemetry.go

```diff
@@ -0,0 +1,121 @@
+	if t.shouldInstallTelemetry {
+		log.Info("starting telemetry for:", "component", t.name)
+		errCh := make(chan error, 1)
+		// Initialize metrics before we start the corresponding controller.
+		// There is a tiny window to miss events here, but should be ok for telemetry purposes.
+		t.initializeAndUpdateMetrics(api)
+		// Start our controller in a goroutine so that we do not block.
+		go func() {
+			defer close(errCh)
+			// Block until our controller manager is elected leader. We presume our
+			// entire process will terminate if we lose leadership, so we don'telemetry need
+			// to handle that.
+			<-mgr.Elected()
+			// Start our controller. This will block until it is stopped
+			// or the controller returns a starting error.
+			if err := (*t.tc).Start(t.stop); err != nil {
+				log.Error(err, "cannot start telemetry controller for", "component", t.name)
+				errCh <- err
+			}
+		}()
+		select {
+		case err := <-errCh:
+			t.shouldInstallTelemetry = true
+			return err
+		case <-time.After(1 * time.Second): // assume no error, as controller has blocked
+		}
+		t.shouldInstallTelemetry = false
+	}
+	return nil
+}
+
+// TryStopTelemetry stops telemetry per component either Eventing, KnativeKafka or Serving
+// When called it assumes that we are reconciling a deletion event.
+func (t *Telemetry) TryStopTelemetry() {
+	if t == nil {
+		return
+	}
+	if !t.shouldInstallTelemetry {
+		log.Info("stopping telemetry for:", "component", t.name)
+		// Stop the telemetry controller
+		close(t.stop)
+		// Can'telemetry use a closed channel
+		t.stop = make(chan struct{})
+		t.shouldInstallTelemetry = true
+	}
+}
+
+// initializeAndUpdateMetrics is used for taking a global snapshot of metrics
+// before we start a telemetry controller. Cost should be low since we are fetching from cache.
+func (t *Telemetry) initializeAndUpdateMetrics(api client.Client) {
+	if t == nil {
+		return
+	}
+	switch t.name {
+	case "eventing":
+		pingSourceG = serverlessTelemetryG.WithLabelValues("source_ping")
+		apiServerSourceG = serverlessTelemetryG.WithLabelValues("source_apiserver")
+		sinkBindingSourceG = serverlessTelemetryG.WithLabelValues("source_sinkbinding")
+		updateMetricFor(&eventingsourcesv1beta1.ApiServerSource{}, api)
+		updateMetricFor(&eventingsourcesv1beta1.PingSource{}, api)
+		updateMetricFor(&eventingsourcesv1beta1.SinkBinding{}, api)
+	case "knativeKafka":
+		kafkaSourceG = serverlessTelemetryG.WithLabelValues("source_kafka")
+		updateMetricFor(&kafkasourcev1beta1.KafkaSource{}, api)
+	case "serving":
+		serviceG = serverlessTelemetryG.WithLabelValues("service")
+		routeG = serverlessTelemetryG.WithLabelValues("route")
+		revisionG = serverlessTelemetryG.WithLabelValues("revision")
+		configurationG = serverlessTelemetryG.WithLabelValues("configuration")
+		updateMetricFor(&servingv1.Service{}, api)
+		updateMetricFor(&servingv1.Route{}, api)
+		updateMetricFor(&servingv1.Revision{}, api)
+		updateMetricFor(&servingv1.Configuration{}, api)
+	}
```

Comment:
Are these lists redundant with the constructor?

I.e. could we store the objects passed and then do:

```golang
for _, obj := t.objects {
  updateMetricFor(obj, api)
}
```